fraud/bad_encoding_test.go

```diff
@@ -0,0 +1,62 @@
+package fraud
+
+import (
+	"context"
+	"errors"
+	"testing"
+
+	format "github.com/ipfs/go-ipld-format"
+	mdutils "github.com/ipfs/go-merkledag/test"
+	"github.com/stretchr/testify/require"
+	"github.com/tendermint/tendermint/pkg/consts"
+	"github.com/tendermint/tendermint/pkg/da"
+	"github.com/tendermint/tendermint/pkg/wrapper"
+
+	"github.com/celestiaorg/celestia-node/ipld"
+	"github.com/celestiaorg/celestia-node/service/header"
+	"github.com/celestiaorg/nmt"
+	"github.com/celestiaorg/rsmt2d"
+)
+
+func TestFraudProofValidation(t *testing.T) {
+	dag := mdutils.Mock()
+	eds := ipld.RandEDS(t, 2)
+	size := eds.Width()
+
+	shares := flatten(eds)
+	copy(shares[3][8:], shares[4][8:])
+	batchAdder := ipld.NewNmtNodeAdder(
+		context.Background(),
+		format.NewBatch(context.Background(), dag, format.MaxSizeBatchOption(int(size)*2)),
+	)
+	tree := wrapper.NewErasuredNamespacedMerkleTree(uint64(size/2), nmt.NodeVisitor(batchAdder.Visit))
+	attackerEDS, _ := rsmt2d.ImportExtendedDataSquare(shares, consts.DefaultCodec(), tree.Constructor)
+	err := batchAdder.Commit()
+	require.NoError(t, err)
+
+	da := da.NewDataAvailabilityHeader(attackerEDS)
+	r := ipld.NewRetriever(dag, consts.DefaultCodec())
+	_, err = r.Retrieve(context.Background(), &da)
+	var errByz *ipld.ErrByzantine
+	require.True(t, errors.As(err, &errByz))
+
+	dah := &header.ExtendedHeader{DAH: &da}
+
+	p := CreateBadEncodingProof(uint64(dah.Height), errByz)
+	err = p.Validate(dah)
+	require.NoError(t, err)
+
+}
+
+func flatten(eds *rsmt2d.ExtendedDataSquare) [][]byte {
+	flattenedEDSSize := eds.Width() * eds.Width()
+	out := make([][]byte, flattenedEDSSize)
+	count := 0
+	for i := uint(0); i < eds.Width(); i++ {
+		for _, share := range eds.Row(i) {
+			out[count] = share
+			count++
+		}
+	}
+	return out
+}
```

Comment:
nit:There is a helper in ipld called ExtractEDS, which does exactly this

Reply:
It is called ExtractODS lol, which is a typo you can also fix

Reply:
Actually, it is called ODS because it extracts the Original square, so this helper is unusable in this case. What we can do thought, is to define another ExtractEDS which does the equivalent of the above functions which we will then reuse everywhere.

Reply:
Not a typo, because it returns original data square. And flatted - full data square